projects/fluid_simulation/main.cpp

```diff
@@ -0,0 +1,146 @@
+    const ppx::grfx::Api kApi = ppx::grfx::API_VK_1_1;
+#endif
+
+    settings.appName               = "fluid_simulation";
+    settings.enableImGui           = false;
+    settings.grfx.api              = kApi;
+    settings.grfx.enableDebug      = true;
+    settings.allowThirdPartyAssets = true;
+
+#if defined(USE_DXIL)
+    settings.grfx.enableDXIL = true;
+#endif
+}
+
+void ProjApp::Setup()
+{
+    // Create the main simulation driver.
+    mSim = std::make_unique<FluidSimulation>(this);
+
+    // Generate the initial screen with random splashes of color.
+    mSim->GenerateInitialSplat();
+}
+
+void ProjApp::Render()
+{
+    PerFrame& frame = mSim->GetFrame(0);
+
+    ppx::grfx::SwapchainPtr swapchain = GetSwapchain();
+
+    uint32_t imageIndex = UINT32_MAX;
+    PPX_CHECKED_CALL(swapchain->AcquireNextImage(UINT64_MAX, frame.imageAcquiredSemaphore, frame.imageAcquiredFence, &imageIndex));
+
+    // Wait for and reset the image-acquired fence.
+    PPX_CHECKED_CALL(frame.imageAcquiredFence->WaitAndReset());
+
+    // Wait for and reset the render-complete fence.
+    PPX_CHECKED_CALL(frame.renderCompleteFence->WaitAndReset());
+
+    // Build the command buffer.
+    PPX_CHECKED_CALL(frame.cmd->Begin());
+    {
+        // Dispatch all the scheduled compute shaders.
+        mSim->DispatchComputeShaders(frame);
+
+        ppx::grfx::RenderPassPtr renderPass = GetSwapchain()->GetRenderPass(imageIndex);
+        PPX_ASSERT_MSG(!renderPass.IsNull(), "render pass object is null");
+        frame.cmd->SetScissors(renderPass->GetScissor());
+        frame.cmd->SetViewports(renderPass->GetViewport());
+
+        frame.cmd->TransitionImageLayout(renderPass->GetRenderTargetImage(0), PPX_ALL_SUBRESOURCES, ppx::grfx::RESOURCE_STATE_PRESENT, ppx::grfx::RESOURCE_STATE_RENDER_TARGET);
+        frame.cmd->BeginRenderPass(renderPass);
+        {
+            mSim->DispatchGraphicShaders(frame, imageIndex);
+
+            // Draw ImGui.
+            DrawDebugInfo();
+            DrawImGui(frame.cmd);
+        }
+        frame.cmd->EndRenderPass();
+        frame.cmd->TransitionImageLayout(renderPass->GetRenderTargetImage(0), PPX_ALL_SUBRESOURCES, ppx::grfx::RESOURCE_STATE_RENDER_TARGET, ppx::grfx::RESOURCE_STATE_PRESENT);
+    }
+    PPX_CHECKED_CALL(frame.cmd->End());
+
+    // Submit the command buffer.
+    ppx::grfx::SubmitInfo submitInfo = {};
+    submitInfo.commandBufferCount    = 1;
+    submitInfo.ppCommandBuffers      = &frame.cmd;
+    submitInfo.waitSemaphoreCount    = 1;
+    submitInfo.ppWaitSemaphores      = &frame.imageAcquiredSemaphore;
+    submitInfo.signalSemaphoreCount  = 1;
+    submitInfo.ppSignalSemaphores    = &frame.renderCompleteSemaphore;
+    submitInfo.pFence                = frame.renderCompleteFence;
+    PPX_CHECKED_CALL(GetGraphicsQueue()->Submit(&submitInfo));
+
+    // Present and mark.
+    PPX_CHECKED_CALL(swapchain->Present(imageIndex, 1, &frame.renderCompleteSemaphore));
+
+    mSim->FreeComputeShaders();
```

Comment:
This is by design.  I want to free up the resources taken by compute shader on every run so I don't run out of descriptors, buffers, etc.  Each iteration of the simulation will allocate new descriptors and buffers.  Alternately, I'd have to keep track of them, but I figured "why not use the API facilities".  If this creates too much overhead, I can re-implement to use my own pool of descriptors, buffers, etc.

Reply:
It depends on the driver and API, but in DX, it can become expensive to create descriptors every frame. That said, you're not creating enough to be such a big problem, so it's definitely fine for now. 

But maybe `FreeComputeShadersResources` is a better name? When first reading this I thought you were unloading the shaders/pipelines.